src/main/java/org/elasticsearch/index/IndexService.java

```
@@ -368,78 +379,70 @@ public synchronized void removeShard(int shardId, String reason) throws Elastics
+                    MergePolicyProvider.class,
+                    IndexShardGatewayService.class,
+                    Translog.class,
+                    PercolatorQueriesRegistry.class);
+
+            // call this before we close the store, so we can release resources for it
+            indicesLifecycle.afterIndexShardClosed(sId, indexShard);
+        } finally {
             try {
-                shardInjector.getInstance(Engine.class).close();
-            } catch (Throwable e) {
-                logger.debug("[{}] failed to close engine", e, shardId);
-                // ignore
-            }
-            try {
-                shardInjector.getInstance(MergeSchedulerProvider.class).close();
-            } catch (Throwable e) {
-                logger.debug("[{}] failed to close merge policy scheduler", e, shardId);
-                // ignore
-            }
-            try {
-                shardInjector.getInstance(MergePolicyProvider.class).close();
-            } catch (Throwable e) {
-                logger.debug("[{}] failed to close merge policy provider", e, shardId);
-                // ignore
-            }
-            try {
-                shardInjector.getInstance(IndexShardGatewayService.class).close();
-            } catch (Throwable e) {
-                logger.debug("[{}] failed to close index shard gateway", e, shardId);
-                // ignore
-            }
-            try {
-                // now we can close the translog
-                shardInjector.getInstance(Translog.class).close();
+                shardInjector.getInstance(Store.class).close();
             } catch (Throwable e) {
-                logger.debug("[{}] failed to close translog", e, shardId);
-                // ignore
+                logger.warn("[{}] failed to close store on shard deletion", e, shardId);
             }
+        }
+    }
+
+    private void close(ShardId shardId, Injector shardInjector, Class<? extends Closeable>... toClose) {
```

Also, can you add javadoc for this (just something mentioning that this swallows exceptions)